### PR TITLE
HPCC-23737 DFUPlus files should not automatically expire with default expiry days if no expireDays parameter present.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Change expireDays default value from '0' to '-1'
Improve FileSpray '@expireDays' attribute handling
Extend spray_expire_test.ecl' to execute dfuplus with and without expire days parameter.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested manually.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: dali/ft/filecopy.cpp [hunk 1]
```diff
@@ -3358,10 +3358,11 @@ void FileSprayer::updateTargetProperties()
             curHistory->addPropTree("Origin",newRecord.getClear());
         }
 
-        int expireDays = options->getPropInt("@expireDays");
-        if (expireDays != -1)
+        if (options->hasProp("@expireDays"))
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp:3361]: simpler to change add -1 as default to existing code, i.e. change code to:
```
    int expireDays = options->getPropInt("@expireDays", -1);
    if (expireDays != -1)
```

File: dali/dfuplus/dfuplus.cpp [hunk 2]
```diff
@@ -895,6 +895,9 @@ int CDfuPlusHelper::copy()
     else if(globals->hasProp("noRecover"))
         req->setNorecover(globals->getPropBool("noRecover", false));
 
+    if(globals->hasProp("expireDays"))
+        req->setExpireDays(globals->getPropInt("expireDays"));
```
Comment (jakesmith) [inline dali/dfuplus/dfuplus.cpp:899]: looks right, but this is a slightly different issue right?
In that it was impossible to set expireDays via a dfuplus copy command line before this change?

File: esp/scm/ws_dfu.ecm [hunk 3]
```diff
@@ -56,13 +56,13 @@ ESPenum SecAccessType : string
 
 /* The FileAccessRole differentiates what type of info is returned in the response
  * "Token" will generally be used to refresh an expired token.
- * "Engine" may not be required long term, but allows the service to tailor and hopefully optimize the reply with info. the engines need. 
+ * "Engine" may not be required long term, but allows the service to tailor and hopefully optimize the reply with info. the engines need.
```
Comment (jakesmith) [inline esp/scm/ws_dfu.ecm:59]: can't actually see what's changed here or why? white spacing?

File: esp/scm/ws_dfu.ecm [hunk 4]
```diff
@@ -56,13 +56,13 @@ ESPenum SecAccessType : string
...
  */
 ESPenum FileAccessRole : string
 {
     Token("Token"),       // Request for security token only, no plain text file information will be return in the response. May be used to refresh an expired token.
-    Engine("Engine"),     // For internal HPCC engines, plain text file meta info may differ from external clients (for now at least). 
-    External("External"), // For external clients (e.g. Spark), plain text less verbose than legacy format needed by engines (e.g. only needs locations, # parts only). 
+    Engine("Engine"),     // For internal HPCC engines, plain text file meta info may differ from external clients (for now at least).
+    External("External"), // For external clients (e.g. Spark), plain text less verbose than legacy format needed by engines (e.g. only needs locations, # parts only).
```
Comment (jakesmith) [inline esp/scm/ws_dfu.ecm:65]: can't actually see what's changed here or why? white spacing?

File: esp/scm/ws_fs.ecm [hunk 5]
```diff
@@ -93,7 +93,7 @@ ESPStruct [nil_remove] DFUWorkunit
     [min_ver("1.09")] bool recordStructurePresent(false);
     [min_ver("1.10")] bool quotedTerminator(true);
     [min_ver("1.12")] bool preserveCompression(true);
-    [min_ver("1.14")] int expireDays;
+    [min_ver("1.14")] int expireDays(-1);
```
Comment (jakesmith) [inline esp/scm/ws_fs.ecm:96]: not sure this is right. Should be left unset I think.
The esp code should check if set and only use it if it is.
It looks like some of the code does, e.g. in CFileSprayEx::onCopy
```
        if (!req.getExpireDays_isNull())
            wuOptions->setExpireDays(req.getExpireDays());
```

but other code doesn't, e.g. in CFileSprayEx::onSprayFixed

File: esp/scm/ws_fs.ecm [hunk 6]
```diff
@@ -330,7 +330,7 @@ ESPrequest [nil_remove] SprayFixed
     [min_ver("1.09")] bool recordStructurePresent(false);
 
     [min_ver("1.10")] bool quotedTerminator(true);
-    [min_ver("1.14")] int expireDays;
+    [min_ver("1.14")] int expireDays(-1);
```
Comment (jakesmith) [inline esp/scm/ws_fs.ecm:333]: see previous comment.

File: esp/scm/ws_fs.ecm [hunk 7]
```diff
@@ -383,7 +383,7 @@ ESPrequest [nil_remove] SprayVariable
     [min_ver("1.10")] bool quotedTerminator(true);
     [min_ver("1.11")] string sourceRowPath;
     [min_ver("1.11")] bool isJSON(false);
-    [min_ver("1.14")] int expireDays;
+    [min_ver("1.14")] int expireDays(-1);
```
Comment (jakesmith) [inline esp/scm/ws_fs.ecm:386]: see previous comment.

File: esp/scm/ws_fs.ecm [hunk 8]
```diff
@@ -476,7 +476,7 @@ ESPrequest [nil_remove] Copy
 
     [min_ver("1.12")] bool preserveCompression(true);
     [min_ver("1.14")] string DFUServerQueue;
-    [min_ver("1.17")] int ExpireDays;
+    [min_ver("1.17")] int ExpireDays(-1);
```
Comment (jakesmith) [inline esp/scm/ws_fs.ecm:479]: see previous comment.

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@AttilaVamos  - a few comments/questions.

File: dali/dfuplus/dfuplus.cpp (hunk 2 above)
Comment (AttilaVamos) [reply dali/dfuplus/dfuplus.cpp:899]: Changed

File: esp/scm/ws_dfu.ecm (hunk 3 above)
Comment (AttilaVamos) [reply esp/scm/ws_dfu.ecm:59]: Whitespaces only

File: esp/scm/ws_dfu.ecm (hunk 4 above)
Comment (AttilaVamos) [reply esp/scm/ws_dfu.ecm:65]: White spaces only

File: esp/scm/ws_fs.ecm (hunk 5 above)
Comment (AttilaVamos) [reply esp/scm/ws_fs.ecm:96]: getExpireDays_isNull() is added to onSprayXXXX() methods.

Comment (AttilaVamos) [reply esp/scm/ws_fs.ecm:96]: -1 removed

File: esp/scm/ws_fs.ecm (hunk 6 above)
Comment (AttilaVamos) [reply esp/scm/ws_fs.ecm:333]: -1 removed

File: esp/scm/ws_fs.ecm (hunk 7 above)
Comment (AttilaVamos) [reply esp/scm/ws_fs.ecm:386]: -1 removed

File: esp/scm/ws_fs.ecm (hunk 8 above)
Comment (AttilaVamos) [reply esp/scm/ws_fs.ecm:479]: -1 removed

**Comment by AttilaVamos** [comment]
@jakesmith updated, tested locally with DFUPlus and EclWatch. Please review.

**Comment by jakesmith** [comment]
@AttilaVamos - looks good.

**Comment by jakesmith** [comment]
@AttilaVamos - can you remind me (and add a comment to the JIRA), when this issue was introduced? Was it when expireDays in sprays was added? Which version was that?

**Comment by AttilaVamos** [comment]
@jakesmith  the answer for your question has been added as a comment to the JIRA.

**Comment by AttilaVamos** [comment]
@richardkchapman please merge.

**Pull request merged**